src/Orleans.Runtime/GrainDirectory/GrainDirectoryHandoffManager.cs

```diff
@@ -271,6 +263,137 @@ internal void ProcessSiloAddEvent(SiloAddress addedSilo)
             }
         }
 
+        private async Task ProcessAddedSiloAsync(SiloAddress addedSilo, List<ActivationAddress> splitPartListSingle, List<ActivationAddress> splitPartListMulti)
+        {
+            if (!this.localDirectory.Running) return;
+
+            if (this.siloStatusOracle.GetApproximateSiloStatus(addedSilo) == SiloStatus.Active)
+            {
+                if (splitPartListSingle.Count > 0)
+                {
+                    if (logger.IsEnabled(LogLevel.Debug)) logger.Debug("Sending " + splitPartListSingle.Count + " single activation entries to " + addedSilo);
+                }
+
+                if (splitPartListMulti.Count > 0)
+                {
+                    if (logger.IsEnabled(LogLevel.Debug)) logger.Debug("Sending " + splitPartListMulti.Count + " entries to " + addedSilo);
+                }
+
+                await localDirectory.GetDirectoryReference(addedSilo).AcceptSplitPartition(splitPartListSingle, splitPartListMulti);
+            }
+            else
+            {
+                if (logger.IsEnabled(LogLevel.Debug)) logger.Debug("Silo " + addedSilo + " is no longer active and therefore cannot receive this partition split");
+            }
```

Comment:
So if we take long enough to successfully process the added silo that the silo is no longer active then we remove the grains from our directory without deactivating them.

I think it's an unsafe change of behavior. Should we return at this point instead?

Reply:
If so, should be a warning log?

Reply:
Yes you are right we should return instead. But the change of status can also happen after this check...